Add indirect build tracing docs (#6667)
github/codeql: docs/codeql/codeql-cli/creating-codeql-databases.rst

  This command runs a custom script that contains all of the commands required
  to build the project.

Using indirect build tracing
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

If the CodeQL CLI autobuilders for compiled languages do not work with your CI workflow and you cannot specify

> build commands, you can use indirect build tracing to create a CodeQL database. To use indirect build tracing, your CI system must be able to set custom environment variables for each build action.

Reviewer: I would swap _"cannot specify build commands"_ with _"cannot wrap invocations of build command with `codeql database trace-command`"_. The build commands will be specified somewhere, it's just a question of how practical it would be for the user to put `codeql database trace-command` in front of those build commands.

CodeQL databases are created with indirect build tracing when you run the following command from the checkout root of your project:

::

  codeql database init ... --begin-tracing <database>

You must specify:

- ``<database>``: a path to the new database to be created. This directory will
  be created when you execute the command---you cannot specify an existing
  directory. 
- ``--begin-tracing``: creates scripts that can be used to set up an environment in which build commands will be traced.

You may specify other options for the ``codeql database init`` command as normal.

.. pull-quote:: Note

>     If you are on Windows, set either ``--trace-process-level <number>`` or ``--trace-process-name <parent process name>`` so that the option points to the parent CI process.

Reviewer: _"parent CI process"_ is a little bit vague here. The key point here is that the user needs to select a parent process that is sufficiently high up that all the build steps for the code being analysed will be observed, but ideally not too high up or we will end up tracing other things that we don't care about. (But note that without the environment variables set the tracer doesn't actually _do_ anything, so it's not the end of the world if we end up tracing something we don't care about.)

The ``codeql database init`` command will output a message:

>   ```
>   Created skeleton <database>. This in-progress database is ready to be populated by an extractor.
>   In order to initialise tracing, some environment variables need to be set in the shell your build will run in.
>   A number of scripts to do this have been created in <database>/temp/tracingEnvironment.
>   Please run one of these scripts before invoking your build command.
> 
>   Based on your operating system, we recommend you run: ...
>   ```

Reviewer 2: The test failure is because you've reverted to markdown syntax for this code block (see https://github.com/github/semmle-docs/blob/main/team-docs/Introduction_to_ReStructuredText.md#code-samples).

I _think_ this would work, but can't test it without cloning your fork.

```suggestion
  ::
  
  Created skeleton <database>. This in-progress database is ready to be populated by an extractor.
  In order to initialise tracing, some environment variables need to be set in the shell your build will run in.
  A number of scripts to do this have been created in <database>/temp/tracingEnvironment.
  Please run one of these scripts before invoking your build command.

  Based on your operating system, we recommend you run: ...

```

The ``codeql database init`` command will produce files in ``<database>/temp/tracingEnvironment`` containing environment variables and their values for CodeQL to trace subsequent build steps. These files are named ``start-tracing.{json,sh,bat,ps1}``. Use one of these files with your CI system's mechanism for setting environment variables for future steps. You can:

* Read the JSON file, process it, and print out environment variables in the format expected by your CI system. For example, Azure DevOps expects ``echo "##vso[task.setvariable variable=NAME]VALUE"``.
* Or source the ``sh/bat/ps1`` script so that its variables go into your shell environment.

Build your code and then run the command ``codeql database finalize <database>``.

> You can optionally clean up the environment variables by following the same process as with the ``--begin-tracing`` scripts, except now with ``--end-tracing`` scripts in the same directory.

Reviewer: There is no `--end-tracing` option. I think you mean to say the `end-tracing.{json,sh,bat,ps1}` scripts.

Also note that it's perfectly okay (and in fact preferable) to end the tracing before `codeql database finalize`. We just have no way of telling the user to run the scripts until they do `codeql database finalize` because the CLI doesn't get called until then. The documentation, however, should probably recommend ending tracing as soon as the last build step is complete, to avoid the performance overhead of tracing `codeql database finalize`.

Once you have created a CodeQL database using indirect build tracing, you can work with it like any other CodeQL database. For example, analyze the database, and upload the results if using Code Scanning.

Example of creating a CodeQL database using indirect build tracing
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The following example shows how you could use indirect build tracing in an Azure DevOps pipeline to create a CodeQL database::

   steps:

>        # Download the CodeQL CLI and query packs...

Reviewer 3: Cancel the open source

Reviewer 3: Secure the source

       - task: PowerShell@1
          displayName: Set CodeQL environment variables
          inputs:
             targetType: inline
             script: >
                $json = Get-Content $(System.DefaultWorkingDirectory)/db/temp/tracingEnvironment/start-tracing.json | ConvertFrom-Json
                $json.PSObject.Properties | ForEach-Object {
                    $template = "##vso[task.setvariable variable="
                    $template += $_.Name
                    $template += "]"
                    $template += $_.Value
                    echo "$template"
                }

       # Execute the pre-defined build step. Note the `msbuildArgs` variable.
       - task: VSBuild@1
           inputs:
             solution: '**/*.sln'
             # Disable MSBuild shared compilation for C# builds.
             msbuildArgs: /p:OutDir=$(Build.ArtifactStagingDirectory) /p:UseSharedCompilation=false
             platform: Any CPU
             configuration: Release
             # Execute a clean build, in order to remove any existing build artifacts prior to the build.
             clean: True
          displayName: Visual Studio Build

>        - task: CmdLine@2

Reviewer: There should probably be a step before this to end tracing - it would look exactly like the start tracing step above but use the `end-tracing` scripts instead. Of course, ending tracing is not strictly required (unless you happen to build something else that you don't want the extractor to see later on in the pipeline) but I think we should be encouraging users to do it.

Author: Thanks. I added another step to this example in the most recent commit. Would you be able to review it to make sure it is accurate? 😄

Reviewer: Seems reasonable to me! @adityasharad would you mind taking a look too? I believe you wrote the original example.

Reviewer 4: End-tracing step looks sensible.

You must specify:

- ``<database>``: a path to the new database to be created. This directory will
  be created when you execute the command---you cannot specify an existing
  directory. 
- ``--begin-tracing``: creates scripts that can be used to set up an environment in which build commands will be traced.

You may specify other options for the ``codeql database init`` command as normal.

.. pull-quote:: Note

    If the build runs on Windows, you must set either ``--trace-process-level <number>`` or ``--trace-process-name <parent process name>`` so that the option points to a parent CI process that will observe all build steps for the code being analyzed.

The ``codeql database init`` command will output a message::

  Created skeleton <database>. This in-progress database is ready to be populated by an extractor.
  In order to initialise tracing, some environment variables need to be set in the shell your build will run in.
  A number of scripts to do this have been created in <database>/temp/tracingEnvironment.
  Please run one of these scripts before invoking your build command.

  Based on your operating system, we recommend you run: ...

> The ``codeql database init`` command creates ``<database>/temp/tracingEnvironment`` with files that contain environment variables and values that will enable CodeQL to trace a sequence of build steps. These files are named ``start-tracing.{json,sh,bat,ps1}``. Use one of these files with your CI system's mechanism for setting environment variables for future steps. You can:

Reviewer 4: This is a good explanation.

* Read the JSON file, process it, and print out environment variables in the format expected by your CI system. For example, Azure DevOps expects ``echo "##vso[task.setvariable variable=NAME]VALUE"``.

> * Or source the appropriate ``start-tracing`` script to set the CodeQL variables in the shell environment of the CI system.

Reviewer 4: Customers often run into trouble when they attempt to source the script but their CI system doesn't persist the environment variables for future steps. If you can think of a good way to phrase it, emphasise here that sourcing the script will only work if the environment is persisted. The overall goal is to set the environment variables.

> Build your code and then run the command ``codeql database finalize <database>``. Optionally, after building the code, unset the environment variables using an ``end-tracing.{json,sh,bat,ps1}`` script from the directory where the ``start-tracing`` scripts are stored.

Reviewer 4: Might want to switch the order around like in your example: build, end tracing, finalize.

> Once you have created a CodeQL database using indirect build tracing, you can work with it like any other CodeQL database. For example, analyze the database, and upload the results to GitHub if you use code scanning.

Reviewer 4: Optional: link to the docs on analyzing and uploading?

Example of creating a CodeQL database using indirect build tracing
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The following example shows how you could use indirect build tracing in an Azure DevOps pipeline to create a CodeQL database::

   steps:
       # Download the CodeQL CLI and query packs...
       # Check out the repository ...

       # Run any pre-build tasks, for example, restore NuGet dependencies...

       # Initialize the CodeQL database.
       # In this example, the CodeQL CLI has been downloaded and placed on the PATH.

>        # If no language is specified, a GitHub Apps or personal access token must be passed through stdin.
>        # to autodetect the language.

Reviewer 2: This comment is from the original and I overlooked line 288 in my earlier review. I think a copy/paste operation went slightly awry here. Based on [`--language`](https://codeql.github.com/docs/codeql-cli/manual/database-create/) I suspect this ought to be:

```suggestion
       # If you omit `--language`, the CLI will call the GitHub API for language data.
       # This will fail unless a GitHub Apps or personal access token is available in the
       # environment variable GITHUB_TOKEN or passed through stdin using `--github-auth-stdin`.
```

However, we don't mention omitting the `--language` option anywhere else in this article, so I wonder if we really want to introduce it here.

Reviewer: I agree with Felicity here - I think we want to keep the example limited to talking about indirect tracing so I would propose that we do not need to mention `--language` all and can just removed the rest of this comment after _"In this example, the CodeQL CLI has been downloaded and placed on the PATH."_.

Documenting exactly how `codeql database init` behaves belongs in the documentation for that command, I think.

       - task: CmdLine@1
          displayName: Initialize CodeQL database
         inputs:
             # Assumes the source code is checked out to the current working directory.
             # Creates a database at `<current working directory>/db`.
             # Running on Windows, so specifies a trace process level.

>              script: "codeql database init --language csharp --trace-process-level 3 --source-root --begin-tracing db"

Reviewer 4: Discussed this with @edoardopirovano. For AzureDevOps we think that `--trace-process-level 3` happens to work, but a more principled suggestion would be to use `--trace-process-name Agent.Worker.exe` instead. That looks for the parent CI process by name instead of its position in the process tree.

Reviewer 4: This line is also missing the value of the `--source-root` argument. That should be the path to the source code, so `--source-root .` if we assume the working directory contains the source code.

             script: >
                $json = Get-Content $(System.DefaultWorkingDirectory)/db/temp/tracingEnvironment/start-tracing.json | ConvertFrom-Json
                $json.PSObject.Properties | ForEach-Object {
                    $template = "##vso[task.setvariable variable="
                    $template += $_.Name
                    $template += "]"
                    $template += $_.Value
                    echo "$template"
                }

       # Execute the pre-defined build step. Note the `msbuildArgs` variable.
       - task: VSBuild@1
           inputs:
             solution: '**/*.sln'
             # Disable MSBuild shared compilation for C# builds.
             msbuildArgs: /p:OutDir=$(Build.ArtifactStagingDirectory) /p:UseSharedCompilation=false
             platform: Any CPU
             configuration: Release
             # Execute a clean build, in order to remove any existing build artifacts prior to the build.
             clean: True
          displayName: Visual Studio Build

       # Read and set the generated environment variables to end build tracing. This is done in PowerShell in this example.
       - task: PowerShell@1

>           displayName: Set CodeQL environment variables

Reviewer 4: ```suggestion
          displayName: Clear CodeQL environment variables
```

  This command runs a custom script that contains all of the commands required
  to build the project.

Using indirect build tracing
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> If the CodeQL CLI autobuilders for compiled languages do not work with your CI workflow and you cannot wrap invocations of build commands with ``codeql database trace-command``, you can use indirect build tracing to create a CodeQL database. To use indirect build tracing, your CI system must be able to set custom environment variables for each build action.

Reviewer: Yes, `codeql database init` followed by `codeql database trace-command <command>` followed by `codeql database finalize` should be equivalent to `codeql database create --comand=<command>`. The latter is the recommended way to create the DB if you have a single command. The former is the recommended way to create a DB if the build requires multiple commands (and you can't wrap them in a script to make them into a single command) and you can add `codeql database trace-command` in front of each one. The new indirect tracing option addresses the case where:

1. You have multiple build commands.
2. You cannot wrap them with `codeql database trace-command`.

> If the CodeQL CLI autobuilders for compiled languages do not work with your CI workflow and you cannot wrap invocations of build commands with ``codeql database trace-command``, you can use indirect build tracing to create a CodeQL database. To use indirect build tracing, your CI system must be able to set custom environment variables for each build action.

Reviewer 2: Thanks for the extra information. It sounds as if we could do with a short overview giving the options for tracing the build for compiled languages. Possibly this would be better as part of a follow up PR, but I'll leave @ethanpalm to make the call on this.

> If the CodeQL CLI autobuilders for compiled languages do not work with your CI workflow and you cannot wrap invocations of build commands with ``codeql database trace-command``, you can use indirect build tracing to create a CodeQL database. To use indirect build tracing, your CI system must be able to set custom environment variables for each build action.

Author: One of the considerations we're trying to balance is to provide indirect tracing as an option for people who need it without directing people toward it unintentionally. This came up in naming and avoiding calling indirect build tracing an advanced option. It feels to me like indirect build tracing would be better introduced as a troubleshooting option rather than one of several options for tracing the build of compiled languages in general.

Perhaps a line at the beginning of [Creating databases for compiled languages](https://codeql.github.com/docs/codeql-cli/creating-codeql-databases/#creating-databases-for-compiled-languages), after `For compiled languages, CodeQL needs to invoke the required build system to generate a database, therefore the build method must be available to the CLI.` explaining to see below about indirect build tracing if it is relevant to the specific use case. I think this could help direct people who need to use indirect build tracing to the procedure but won't cause people to think they should use indirect build tracing when they don't need to.

> If the CodeQL CLI autobuilders for compiled languages do not work with your CI workflow and you cannot wrap invocations of build commands with ``codeql database trace-command``, you can use indirect build tracing to create a CodeQL database. To use indirect build tracing, your CI system must be able to set custom environment variables for each build action.

Author: After a bit more thinking, I am going to open a separate issue for how we introduce this information because I think there are a few different approaches we can take.

> If the CodeQL CLI autobuilders for compiled languages do not work with your CI workflow and you cannot wrap invocations of build commands with ``codeql database trace-command``, you can use indirect build tracing to create a CodeQL database. To use indirect build tracing, your CI system must be able to set custom environment variables for each build action.

Reviewer 4: I agree with @felicitymay's suggestion about providing a short guide on the options for compiled languages. There are many possibilities, but to me the high level options are:
- Do you use a well-known build system recognised by the CodeQL autobuilders? Use `codeql database create` (without a `--command` argument) to autobuild the code.
- Do you know the build command line? Use `codeql database create ... --command "<build command>"`
  - A variation of this is if you have multiple build command lines, in which case you would use `trace-command` multiple times. I don't _think_ we need to mention that just yet. 
- If neither of the above are suitable, for example if you are using preconfigured build steps from your CI system that do not expose the build command, then use indirect build tracing. Examples of such build steps are the VSBuild and MSBuild tasks in Azure DevOps.

Notably, indirect tracing is not a viable troubleshooting option. Aside from autobuild failing, there's no way to try out a build without it if you don't know the build command.

The ``codeql database init`` command will output a message::

  Created skeleton <database>. This in-progress database is ready to be populated by an extractor.
  In order to initialise tracing, some environment variables need to be set in the shell your build will run in.
  A number of scripts to do this have been created in <database>/temp/tracingEnvironment.
  Please run one of these scripts before invoking your build command.

  Based on your operating system, we recommend you run: ...

The ``codeql database init`` command will produce files in ``<database>/temp/tracingEnvironment`` containing environment variables and their values for CodeQL to trace subsequent build steps. These files are named ``start-tracing.{json,sh,bat,ps1}``. Use one of these files with your CI system's mechanism for setting environment variables for future steps. You can:

* Read the JSON file, process it, and print out environment variables in the format expected by your CI system. For example, Azure DevOps expects ``echo "##vso[task.setvariable variable=NAME]VALUE"``.
* Or source the ``sh/bat/ps1`` script so that its variables go into your shell environment.

> Build your code, end build tracing, and then run the command ``codeql database finalize <database>``.
> 
> You can optionally clean up the environment variables by following the same process as with the ``--begin-tracing`` scripts, except now with ``end-tracing.{json,sh,bat,ps1}`` scripts in the same directory.

Reviewer 2: On line 270, we haven't told them how to "end build tracing" yet, I think we do this on the line below. So maybe this would be clearer?

```suggestion
Build your code and then run the command ``codeql database finalize <database>``. Optionally, after building the code, unset the environment variables using an ``end-tracing.{json,sh,bat,ps1}`` script from the directory where the ``start-tracing`` scripts are stored.
```

> Once you have created a CodeQL database using indirect build tracing, you can work with it like any other CodeQL database. For example, analyze the database, and upload the results if using Code Scanning.

Reviewer 2: ```suggestion
Once you have created a CodeQL database using indirect build tracing, you can work with it like any other CodeQL database. For example, analyze the database, and upload the results to GitHub if you use code scanning.
```

Example of creating a CodeQL database using indirect build tracing
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The following example shows how you could use indirect build tracing in an Azure DevOps pipeline to create a CodeQL database::

   steps:
       # Download the CodeQL CLI and query packs...
       # Check out the repository ...

>        # Tasks prior to executing the build, e.g. restore NuGet dependencies...

Reviewer 2: ```suggestion
       # Run any pre-build tasks, for example, restore NuGet dependencies...
```

       # Initialize the CodeQL database.
       # In this example, the CodeQL CLI has been downloaded and placed on the PATH.

>        # If no language is specified, a GitHub Apps or personal access token must be passed through stdin

Reviewer 2: We seem mostly to use periods or ellipses for these comments, so adding one here.

```suggestion
       # If no language is specified, a GitHub Apps or personal access token must be passed through stdin.
```

       # to autodetect the language.
       - task: CmdLine@1
          displayName: Initialize CodeQL database
         inputs:
             # Assumes the source code is checked out to the current working directory.

>              # Creates a database at `<current working directory>/db`

Reviewer 2: I think we need to mention that this is running on a Windows server:

```suggestion
             # Creates a database at `<current working directory>/db`.
             # Running on Windows, so specifies a trace process level.
```

             platform: Any CPU
             configuration: Release
             # Execute a clean build, in order to remove any existing build artifacts prior to the build.
             clean: True
          displayName: Visual Studio Build

       # Read and set the generated environment variables to end build tracing. This is done in PowerShell in this example.
       - task: PowerShell@1
          displayName: Set CodeQL environment variables
          inputs:
             targetType: inline
             script: >
                $json = Get-Content $(System.DefaultWorkingDirectory)/db/temp/tracingEnvironment/end-tracing.json | ConvertFrom-Json
                $json.PSObject.Properties | ForEach-Object {
                    $template = "##vso[task.setvariable variable="
                    $template += $_.Name
                    $template += "]"
                    $template += $_.Value
                    echo "$template"
                }

       - task: CmdLine@2
          displayName: Finalize CodeQL database
          inputs:
             script: 'codeql database finalize db'

>        # Other tasks go here,
>        # e.g. `codeql database analyze`
>        # and `codeql github upload-results` ...

Reviewer 2: ```suggestion
       # Other tasks go here, for example:
       # `codeql database analyze`
       # then `codeql github upload-results` ...
```
